File: py/0039_CombinationSum.py
# 39. Combination Sum

from typing import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        curr = []

        def backtrack(i: int, total: int):
            if total == target:
                result.append(curr.copy())
                return
            if i == len(candidates) or total > target:
                return

            curr.append(candidates[i])
            backtrack(i, total + candidates[i])
            curr.pop()
            backtrack(i + 1, total)

        backtrack(0, 0)
        return result


if __name__ == "__main__":
    puzzles = [
        ([2, 3, 6, 7], 9),
        ([2, 3, 5], 8),
        ([2], 1)
    ]
    for puzzle in puzzles:
        print(Solution().combinationSum(*puzzle))

File: go/0039/0039_CombinationSum.go
// 39. Combination Sum

package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"leetcode/utils"
	"os"
	"slices"

	"github.com/samber/lo"
)

func backtrack(candidates []int, target int, i int, total int, result *[][]int, path *[]int) {
	// either add the current value or skip to the next
	// we do not need to consider add and take next since it happens in the next recursion
	if total == target {
		*result = append(*result, slices.Clone(*path))
	}
	if total >= target || i == len(candidates) {
		return
	}

	backtrack(candidates, target, i+1, total, result, path) // skip current
	*path = append(*path, candidates[i])
	backtrack(candidates, target, i, total+candidates[i], result, path) // add current value
	*path = (*path)[:len(*path)-1]
}

func combinationSum(candidates []int, target int) [][]int {
	var result [][]int
	var path []int

	slices.Sort(candidates)
	backtrack(candidates, target, 0, 0, &result, &path)

	return result
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	var lines []string
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	const testSize int = 2

	for _, test := range lo.Chunk(lines, testSize) {
		var candidates []int
		json.Unmarshal([]byte(test[0]), &candidates)
		target := utils.ParseInt(test[1])
		result := combinationSum(candidates, target)
		fmt.Println(result)
	}
}
